tornado/gen.py

```diff
@@ -244,7 +245,8 @@ def coroutine(func, replace_callback=True):
     """
     return _make_coroutine_wrapper(func, replace_callback=True)
 
-
+# Ties lifetime of futures to their runners. Github Issue #1769
+_futures_to_runners = weakref.WeakKeyDictionary()
```

Comment:
Add a blank line after this definition, and expand the comment. Something like:
Unlike regular stack frames, idle coroutines are not anchored to a garbage-collection root. Object references point in the opposite direction from what you'd expect, with child frames holding references to the parent frames that they will awaken when completed, but there is not necessarily anything keeping the child frames alive (In most cases the innermost child frame will be registered as a callback on the IOLoop, but other patterns are possible especially when weak references are used). This map provides hard references in the other direction, so that as long as the Future returned by a coroutine is referenced (typically by the parent's Runner), this coroutine's Runner will be kept alive.